CRM/Consentactivity/Service.php

```diff
@@ -234,21 +251,226 @@ public static function tagExists(int $tagId): bool
+     * Also sets the privacy flags.
+     *
+     * @param int $contactId
+     */
+    public static function anonymizeContact(int $contactId): void
+    {
+        foreach (self::CONTACT_DATA_ENTITIES as $entity) {
+            $numberOfEntities = $entity::get(false)
+                ->addWhere('contact_id', '=', $contactId)
+                ->selectRowCount()
+                ->execute();
+            if (count($numberOfEntities)) {
+                $entity::delete(false)
+                    ->addWhere('contact_id', '=', $contactId)
+                    ->setLimit(count($numberOfEntities))
+                    ->execute();
+            }
+        }
+        $contactFieldsToDelete = [
+            'first_name', 'last_name', 'middle_name', 'display_name',
+            'email_greeting_display', 'postal_greeting_display',
+            'addressee_display', 'nick_name', 'sort_name',
+            'external_identifier', 'image_url', 'api_key', 'birth_date',
+            'deceased_date', 'employer_id', 'job_title', 'gender_id',
+        ];
+        $privacyFieldsToSet = [
+            'do_not_email', 'do_not_phone', 'do_not_mail',
+            'do_not_sms', 'do_not_trade', 'is_opt_out',
+        ];
+        $contactRequest = Contact::update(false)
+            ->addWhere('id', '=', $contactId)
+            ->setLimit(1);
+        foreach ($contactFieldsToDelete as $field) {
+            $contactRequest = $contactRequest->addValue($field, '');
+        }
+        foreach ($privacyFieldsToSet as $field) {
+            $contactRequest = $contactRequest->addValue($field, true);
+        }
+        $contactRequest->execute();
+    }
+    /**
+     * On case of contribution create it adds a consentactivity action
+     * to the contributor contact if it is configured on the settings form.
+     *
+     * @param string $op
+     * @param string $objectName
+     * @param $objectId - the unique identifier for the object.
+     * @param $objectRef - the reference to the object if available.
+     */
+    public static function post(string $op, string $objectName, $objectId, &$objectRef): void
+    {
+        if ($op !== 'create' || $objectName !== 'Contribution' || $objectRef->is_test) {
+            return;
+        }
+        $cfg = new CRM_Consentactivity_Config(E::LONG_NAME);
+        $cfg->load();
+        $config = $cfg->get();
+        if (!$config['consent-after-contribution']) {
+            return;
+        }
+        $receiveDate = date('Y-m-d H:i', strtotime($objectRef->receive_date));
+        $expireDate = date('Y:m-d H:i', strtotime($config['consent-expiration-years'].' years ago'));
+        if ($receiveDate < $expireDate) {
+            return;
+        }
+        $activity = self::createConsentActivityToContact($objectRef->contact_id);
+        if (isset($activity['id'])) {
+            // update activity with sql
+            $sql = "UPDATE civicrm_activity SET created_date = %1, activity_date_time = %1 WHERE id =  %2";
```

Comment:
Why do we need raw sql here?

Reply:
I wasn't able to update the create_date with API. The created_date did not changed.